autogluon/utils/tabular/ml/learner/default_learner.py

```diff
@@ -105,7 +117,120 @@ def _fit(self, X: DataFrame, X_val: DataFrame = None, scheduler_options=None, hy
+                if time_limit < MIN_TIME:
+                    logger.warning(NOT_ENOUGH_TIME_WARNING)
+                    return None
+
+        logger.log(20, "Preprocessing train data together with test data ...")
+        feature_generator_og = self.feature_generator
+        self.feature_generator = feature_generator
+        X, y, X_val, y_val, holdout_frac, num_bagging_folds = self.general_data_processing(X, X_val, holdout_frac, num_bagging_folds)
+        if noval_nobag:  # split  validation set here and avoid including pseudolabels
+            _, holdout_frac, _ = self.adjust_threshold_if_necessary(y.loc['train'], self.threshold, holdout_frac_og, num_bagging_folds)
+            X_train, X_val, y_train, y_val = generate_train_test_split(X.loc['train'], y.loc['train'], problem_type=self.problem_type, test_size=holdout_frac, random_state=self.random_seed)
+            X = pd.concat([X_train, X.loc['test']], ignore_index=True)
+            y = pd.concat([y_train, y.loc['test']], ignore_index=True)
+
+        time_preprocessing_end = time.time()
+        time_preprocessing = time_preprocessing_end - time_pseudolabeling_end
+        logger.log(15, f"Preprocessing train data together with test data runtime = {round(time_preprocessing, 2)}s")
+        if time_limit is not None:
+            time_limit -= time_preprocessing
+            if time_limit < MIN_TIME:
+                logger.warning(NOT_ENOUGH_TIME_WARNING)
+                self.feature_generator = feature_generator_og  # reset
+                return None
+
+        logger.log(20, "Training models (transductively) with unlabeled test data ...")
+        # Store transductively-trained models in separate trainer so they aren't ensembled with other models:
+        trainer = self.trainer_type(
+            path=self.model_context+TRANSDUCTIVE_SUFFIX,
+            problem_type=self.label_cleaner.problem_type_transform,
+            eval_metric=self.eval_metric,
+            stopping_metric=self.stopping_metric,
+            num_classes=self.label_cleaner.num_classes,
+            feature_metadata=self.feature_generator.feature_metadata,
+            low_memory=True,
+            kfolds=num_bagging_folds,
+            n_repeats=num_bagging_sets,
+            stack_ensemble_levels=stack_ensemble_levels,
+            scheduler_options=scheduler_options,
+            time_limit=time_limit,
+            save_data=save_data,
+            save_bagged_folds=save_bagged_folds,
+            random_seed=self.random_seed,
+            verbosity=verbosity
+        )
+        self.inductive_trainer_path = self.trainer_path
+        self.trainer_path = trainer.path
+        trainer.train(X, y, X_val, y_val, hyperparameter_tune=hyperparameter_tune, feature_prune=feature_prune, holdout_frac=holdout_frac,
+                      hyperparameters=hyperparameters, ag_args_fit=ag_args_fit, excluded_model_types=excluded_model_types, name_suffix=TRANSDUCTIVE_SUFFIX)
+        self.save_trainer(trainer=trainer)
```

Comment:
Making predictions with these new models on the test data can cause severe score drop due to the predictions being generated from models which used the pseudo-labels of the test data during training. This is especially visible in log_loss eval_metric, where RF can be extremely overconfident about its training data predictions, leading to a very bad score when a pseudo-label happened to be different than ground truth.

Tested on Adult with eval_metric=log_loss, RF went from 0.307 logloss (without trans) to 0.378 (with trans). In addition, val_score didn't change much, making the trans trained model's val_score misleading (Both had ~0.306 val score).

I believe that for this to be usable, predictions for test rows that were used in pseudo-labeled train data must come from the original models and not the trans models, or they must be trained with soft_class.

Reply:
Pseudo-labeled rows should be excluded from fold validation set when we calculate scores. Otherwise you will extremely over-confident fold metric estimates, which won't generalize.